modules/apigee_edge_apidocs/src/ApiDocSpecFetcher.php

```diff
@@ -0,0 +1,264 @@
+        $specs_definition = $apidoc->getFieldDefinition('spec')->getItemDefinition();
+        $target_dir = $specs_definition->getSetting('file_directory');
+        $uri_scheme = $specs_definition->getSetting('uri_scheme');
+        $destination = "$uri_scheme://$target_dir/";
+
+        try {
+          $this->checkRequirements($destination);
+          $file = file_save_data($data, $destination . $filename, FILE_EXISTS_RENAME);
+
+          if (empty($file)) {
+            throw new \Exception('Could not save API Doc specification file.');
+          }
+        }
+        catch (\Exception $e) {
+          $this->log($apidoc, static::TYPE_ERROR, 'Error while saving API Doc spec file from URL on API Doc ID: %id. Error: %error', [
+            '%id' => $apidoc->id(),
+            '%error' => $e->getMessage(),
+          ], $show_messages);
+          return FALSE;
+        }
+
+        $spec_value = ['target_id' => $file->id()] + $spec_value;
+        $apidoc->set('spec', $spec_value);
+        $apidoc->set('spec_md5', $data_md5);
+        $apidoc->set('fetched_timestamp', time());
+
+        $needs_save = TRUE;
+      }
+    }
+
+    elseif ($apidoc->get('spec_file_source')->value == ApiDocInterface::SPEC_AS_FILE) {
+      $this->log($apidoc, static::TYPE_STATUS, 'API Doc %label is using a file upload as source. Nothing to update.', [
+        '%label' => $apidoc->label(),
+      ], $show_messages);
+      $needs_save = FALSE;
+    }
+
+    return $needs_save;
+  }
+
+  /**
+   * Log a message, and optionally display it on the UI.
+   *
+   * @param \Drupal\apigee_edge_apidocs\Entity\ApiDocInterface $apidoc
+   *   The API Doc entity.
+   * @param string $type
+   *   Type of message.
+   * @param string $message
+   *   The message.
+   * @param array $params
+   *   Optional parameters array.
+   * @param bool $show_messages
+   *   TRUE if message should be displayed to the UI as well.
+   */
+  private function log(ApiDocInterface $apidoc, string $type, string $message, array $params = [], bool $show_messages = TRUE) {
+    switch ($type) {
+      case static::TYPE_ERROR:
+        $this->logger->error($message, $params);
+        break;
+
+      case static::TYPE_STATUS:
+
+      default:
+        $this->logger->info($message, $params);
+    }
+    if ($show_messages) {
+      $this->messenger->addMessage($this->t($message, $params), $type);
+    }
+  }
+
+  /**
+   * Checks requirements for saving of a file spec.
+   *
+   * If a requirement is not fulfilled it throws an exception.
+   *
+   * @param string $destination
+   *   The specification file destination directory, including scheme.
+   *
+   * @throws \Exception
+   */
+  private function checkRequirements(string $destination): void {
+    // If using private filesystem, check that it's been configured.
+    if (strpos($destination, 'private://') === 0 && !$this->isPrivateFileSystemConfigured()) {
+      throw new \Exception('Private filesystem has not been configured.');
```

Comment:
Throw a module-specific exception or something more specific than a generic one so callers could differentiate what went wrong and what was the source of the exception.